ksqldb-rest-app/src/main/java/io/confluent/ksql/rest/server/computation/DistributingExecutor.java

```diff
@@ -172,10 +174,27 @@ public StatementExecutorResponse execute(
           + System.lineSeparator()
           + commandRunnerWarningString);
     }
-    final ConfiguredStatement<?> injected = injectorFactory
-        .apply(executionContext, securityContext.getServiceContext())
-        .inject(statement);
+    final InjectorWithSideEffects injector = InjectorWithSideEffects.wrap(
+            injectorFactory.apply(executionContext, securityContext.getServiceContext()));
+    final ConfiguredStatementWithSideEffects<?> injectedWithSideEffects =
+            injector.injectWithSideEffects(statement);
+    try {
+      return executeInjected(
+            injectedWithSideEffects.getStatement(), statement,
+            executionContext,
+            securityContext);
+    } catch (Exception e) {
+      injector.revertSideEffects(injectedWithSideEffects);
```

Comment:
I'm wondering whether a race condition is possible:
1. Instance A checks the topic, it doesn’t exist, so it creates it
2. Instance B checks the topic, it exists, so it skips creation
3. Instance B fences off instance A
4. Instance A deletes the topic as a result of failure handling
5. Instance B continues execution assuming that the topic exists

If it is possible, we should probably ignore `ProducerFencedException` (assuming that commands always result in this topic being created). This also leaves a very small possibility of not deleting this topic - if the newer producer fails too.

Reply:
As discussed offline: there is a ton of race condition and focusing on the "happy cleanup path" is a good step forward. -- Making it absolutely bullet proof seems to be "impossible" anyway (ie, we would need to change Kafka to get an atomic "create topics and write into command-topic" API)